tsl/src/remote/copy_fetcher.c

```diff
@@ -550,41 +584,13 @@ copy_fetcher_close(DataFetcher *df)
 {
 	CopyFetcher *fetcher = cast_fetcher(CopyFetcher, df);
 
-	/*
-	 * The fetcher state might not be open if the fetcher got initialized but
-	 * never executed due to executor constraints.
-	 */
-	if (fetcher->state.open && fetcher->state.data_req != NULL)
-	{
-		int end_res = PQendcopy(remote_connection_get_pg_conn(fetcher->state.conn));
-		if (end_res != 0)
-		{
-			TSConnectionError err;
-			remote_connection_get_error(fetcher->state.conn, &err);
-			remote_connection_error_elog(&err, ERROR);
-		}
-
-		/*
-		 * Shouldn't have any activity on the connection after we have
-		 * finished COPY. Just double-check.
-		 */
-		PGresult *pgres = PQgetResult(remote_connection_get_pg_conn(fetcher->state.conn));
-		if (pgres)
-		{
-			TSConnectionError err;
-			remote_connection_get_result_error(pgres, &err);
-			if (err.msg == NULL)
-			{
-				err.msg = "internal program error: remaining activity on the data node connection "
-						  "after finishing COPY";
-			}
-			remote_connection_error_elog(&err, ERROR);
-		}
+	/* Check for premature ending of COPY (i.e., before reaching EOF). Note,
+	 * if EOF was reached, we closed already in the main processing loop so no
+	 * need to end_copy() here in that case. */
+	if (!fetcher->state.eof)
+		end_copy_before_eof(fetcher);
```

Comment:
If we already received the file trailer, we shouldn't cancel the query, right? It's finished already.

Reply:
I guess, but I don't really see how it would happen without us also reading EOF. And, if we would cancel at file trailer without receiving EOF, then that's fine too because if we explicitly cancel it is not an error case.

Reply:
I think we shouldn't cancel at file trailer, it's just another kind of eof, so cancelling would be pointless extra work, and it's heavy.

> I guess, but I don't really see how it would happen without us also reading EOF.

Let's add an assertion then. I just want to make sure that `trailer` and `eof` almost always have the same value, and the only exception is `trailer = true and eof = false` while we're making the final `PGgetCopyData` call which is just a sanity check.

Reply:
It is not strictly wrong to cancel at file trailer before reading EOF. This is very unlikely to happen, though, so why optimize for it?

Reply:
Added an extra Assert(), although not sure this is what you were looking for?